wolfssl/src/ssl.rs

```diff
@@ -1100,6 +1127,80 @@ impl<IOCB: IOCallbacks> Session<IOCB> {
             e => unreachable!("{e:?}"),
         }
     }
+
+    /// Enable TLS1.3 key logging for applications
+    #[cfg(feature = "debug")]
+    pub(crate) fn enable_tls13_keylog(&mut self, secret_cb: Tls13SecretCallbacksArg) -> Result<()> {
+        self.secret_cb = Some(Box::new(secret_cb));
+
+        // SAFETY: `secret_cb` is a `Box` pointer so the address is
+        // stable. (The address is the address of the heap allocation
+        // containing the `Tls13Secretcallbacksarg` which is an `Arc`.
+        //
+        // We free `self.ssl` (the `wolfssl_sys::WOLFSSL`) on drop of
+        // `self`, any use of the callback must have stopped before
+        // the drop, since nothing can also be making calls into the session.
+        //
+        // Therefore `secret_cb` here is valid for as long as it needs to be.
+        let secret_cb: &mut Tls13SecretCallbacksArg = self.secret_cb.as_mut().unwrap();
+        let secret_cb: *mut Tls13SecretCallbacksArg = secret_cb as *mut Tls13SecretCallbacksArg;
+        let secret_cb = secret_cb as *mut c_void;
```

Comment:
Strictly speaking the type annotations here might not be needed and this could be a oneliner, but I think the types help clarify what is going on (and help the compiler alert us if something isn't doing what we thing).

I think the first annotation _might_ be needed since it causes auto deref to kick in, without that you'd get `&mut <Box<Arc<dyn ....>>>` not the type we want/need.

Reply:
Looks nice. Thanks!